### PR TITLE
fix(Core/Script): Ouro Spawner & Ouro

### LINKED ISSUE #1689: [RAID] Temple of Ahn'Qiraj (Boss Ouro Doesn't Spawn)
##### SMALL DESCRIPTION:

The optional boss Ouro in AQ40 raid never spawns correctly. So unable to fight it at all.

##### EXPECTED BLIZZLIKE BEHAVIOUR:

https://www.youtube.com/watch?v=uGhW2pvRKA4
https://wow.gamepedia.com/Ouro

##### CURRENT BEHAVIOUR:

You see the boss move around underground but it never emerges from the sand.

##### STEPS TO REPRODUCE THE PROBLEM:

1. .tele aq40
2. fight up to location #8 on the following map:
image
3. You see the boss move around underground and once you are in range, it attacks you, but this is completely wrong. The boss should emerge from the sand and then attack the raid party.

##### BRANCH(ES):

master

##### AC HASH/COMMIT:

9eb6e9a

##### OPERATING SYSTEM:

Win 7/10/Svr 2016

##### MODULES:

N/A

##### OTHER CUSTOMIZATIONS:

N/A

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/72443705-raid-temple-of-ahn-qiraj-boss-ouro-doesn-t-spawn?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by imxari**
Fix(Core/Raid) Ouro Spawner & Ouro

## CHANGES PROPOSED:
-  Ouro Spawner didn't have a script which meant that you couldn't fight Ouro.
-  Ouro timers replaced with EventMap.

## ISSUES ADDRESSED:
- Closes #1689

## TESTS PERFORMED:
- Built the core
- Works in-game

## HOW TO TEST THE CHANGES:

- Build the core using my PR
- Go to the Temple Of Ahn'Qiraj
- Walk near the moving rubble, Ouro will spawn.

## KNOWN ISSUES AND TODO LIST:

- None

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1588445740470602001.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1588445740470602001');
+
+-- Ouro Spawner
+UPDATE `creature_template` SET `unit_flags` = 33685504, `ScriptName` = 'npc_ouro_spawner' WHERE (`entry` = 15957);
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1588445740470602001.sql:4]: for unit flags, please use `unit_flags`=`unitflags`|FLAG so you add the flags and not override the flags

Comment (h0lysp4nk) [reply data/sql/updates/pending_db_world/rev_1588445740470602001.sql]: Okay, fixed

**Comment by Kitzunu** [comment]
Codacy Error:
```
Member variable 'boss_ouroAI::Submerged' is not initialized in the constructor.
boss_ouroAI(Creature* creature) : ScriptedAI(creature) { me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_DISABLE_MOVE); }

**Comment by h0lysp4nk** [comment]
> Codacy Error:
> 
> ```
> Member variable 'boss_ouroAI::Submerged' is not initialized in the constructor.
> boss_ouroAI(Creature* creature) : ScriptedAI(creature) { me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_DISABLE_MOVE); }
> ```

Fixed.

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp [hunk 2]
```diff
@@ -17,11 +17,75 @@ EndScriptData */
...
+    EVENT_CHARGE_TARGET = 5,
+    EVENT_SPAWN = 6
+};
+
+class npc_ouro_spawner : public CreatureScript
+{
+public:
+    npc_ouro_spawner() : CreatureScript("npc_ouro_spawner") { }
+
+    CreatureAI* GetAI(Creature* creature) const
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp:46]: Missing override

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp [hunk 3]
```diff
@@ -72,56 +142,67 @@ class boss_ouro : public CreatureScript
...
-                me->setFaction(14);
-
-                DoCastVictim(SPELL_GROUND_RUPTURE);
-
-                Submerged = false;
-                Submerge_Timer = urand(60000, 120000);
-            } else Back_Timer -= diff;
+                switch (eventId)
+                {
+                    case 0:
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp:154]: Useless

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp (hunk 2 above)
Comment (h0lysp4nk) [reply src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp:46]: Okay, done :)

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp (hunk 3 above)
Comment (h0lysp4nk) [reply src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp]: Removed

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp (hunk 2 above)
Comment (Nefertumm) [reply src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp:46]: Still missing here :P Also the whole tabs got shifted.

Comment (Nefertumm) [reply src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_ouro.cpp:46]: Nevermind, the tabs didn't get shifted.

**Comment by BarbzYHOOL** [comment]
Gonna test this as I am the guy who begged Holyspank to do it :D

**Comment by gengarshadowball** [comment]
Nice, he doest spawn now.
However i had this weird issue.

Whenever I tried to solo it and walked away Ouro teleported into me.

**Comment by h0lysp4nk** [comment]
> Nice, he doest spawn now.
> However i had this weird issue.
> 
> Whenever I tried to solo it and walked away Ouro teleported into me.

@gengarshadowball 

Are you using my PR? I can't reproduce this issue on my local server.

**Comment by gengarshadowball** [comment]
Yes bro, I think you missunderstood because I had a typo. He does spawn with the PR, but he still tries to move around following the player when in combat and since the flag doesn't allow him he just teleports which is weird. I can show a video maybe. I'll post a link when ready!

**Comment by FrancescoBorzi** [comment]
@imxari there are merge conflicts

**Comment by Racknar0** [comment]
I don't know if it's just me, but after applying this commit, the boss doesn't exit, it doesn't let select "focus"

before PR

image


after PR

image

**Comment by FrancescoBorzi** [comment]
@imxari

**Pull request closed without merging**